app/Actions/AlbumAuthorisationProvider.php

```diff
@@ -0,0 +1,553 @@
+	 *
+	 * This method adds the WHERE conditions without any further pre-cautions.
+	 * The method silently assumes that the SELECT clause contains the tables
+	 *
+	 *  - **`base_albums`** and
+	 *  - **`user_base_album`**.
+	 *
+	 * Moreover, the raw OR-clauses are added.
+	 * They are not wrapped into a nesting braces `()`.
+	 *
+	 * @param Builder $query
+	 *
+	 * @return Builder
+	 *
+	 * @throws \InvalidArgumentException
+	 */
+	public function appendAccessibilityConditions(BaseBuilder $query): BaseBuilder
+	{
+		$unlockedAlbumIDs = $this->getUnlockedAlbumIDs();
+		$userID = AccessControl::is_logged_in() ? AccessControl::id() : null;
+
+		$query
+			->orWhere(fn (BaseBuilder $q) => $q
+				->where('base_albums.is_public', '=', true)
+				->whereNull('base_albums.password')
+			)
+			->orWhere(fn (BaseBuilder $q) => $q
+				->where('base_albums.is_public', '=', true)
+				->whereIn('base_albums.id', $unlockedAlbumIDs)
+			);
+		if ($userID !== null) {
+			$query
+				->orWhere('base_albums.owner_id', '=', $userID)
+				->orWhere('user_base_album.user_id', '=', $userID);
+		}
+
+		return $query;
+	}
+
+	/**
+	 * Checks whether the album is accessible by the current user.
+	 *
+	 * For real albums (i.e. albums that are stored in the DB), see
+	 * {@link AlbumAuthorisationProvider::applyAccessibilityFilter()} for a
+	 * specification of the rules when an album is accessible.
+	 * In other cases, the following holds:
+	 *  - the root album is accessible if and only if the user is authenticated
+	 *  - the built-in smart albums are accessible, if
+	 *     - the user is authenticated and is granted the right of uploading, or
+	 *     - the album is the album of recent photos and public by configuration, or
+	 *     - the album is the album of starred photos and public by configuration
+	 *
+	 * Note, this method tries to minimize DB queries and any overhead due
+	 * to hydration of models.
+	 * If an actual instance of a {@link AbstractAlbum} model is passed in,
+	 * then the DB won't be queried at all, because all checks are performed
+	 * on the values of the already hydrated model.
+	 * If an ID is passed, then the method runs a very efficient COUNT
+	 * query on the DB.
+	 * In particular, no {@link Album} nor {@link TagAlbum} model is hydrated
+	 * to avoid any overhead.
+	 *
+	 * Tips for usage:
+	 *  - If you already have a {@link AbstractAlbum} instance, pass that.
+	 *    This is most efficient.
+	 *  - If you do not have a {@link AbstractAlbum} instance, but you will
+	 *    need one later anyway, then use {@link AlbumFactory} to first fetch
+	 *    the album from DB and pass the album.
+	 *    This avoids a second DB query later.
+	 *  - If you do not have a {@link AbstractAlbum} instance, and you won't
+	 *    need one later, simply pass the ID of the album.
+	 *    This avoids the overhead of model hydration.
+	 *
+	 * @param string|int|null $albumID
+	 *
+	 * @return bool
+	 */
+	public function isAccessibleByID($albumID): bool
```

Comment:
In PHP 8, will need to be refined to use Union types. :)
https://stitcher.io/blog/new-in-php-8#union-types-rfc

Reply:
I know, I am looking forward to it.